### DISCUSSION TITLE
Issue associated with restart ArrayVariable

### DISCUSSION THREAD

**Discussion by lingzou** [idaholab/moose#16915] [Q&A General] [answered]
Dear All, I am exploring the ArrayVariable capability, and encountered some issues associated with its restart capability. Here, ArrayVariable is something like:

```
  [./v]
    order = CONSTANT
    family = MONOMIAL
    components = 3
    initial_condition = '12 13 14'
  [../]
```

Please see attached two input files, step-0.i is a simple input file with 'u' a regular variable, and 'v' an ArrayVariable. In 'step-1.i', I try to restart the simulation using step-0 results by giving:

```
[Mesh]
  file = step-0_out.e
[]
```

The error message is attached, which seems to be associated with the restart of the ArrayVariable.

```
map_find() error: key "v" not found in file ../src/systems/system.C on line 1405
Stack frames: 13
0: 0   libmesh_opt.0.dylib                 0x0000000106b0d865 libMesh::print_trace(std::__1::basic_ostream<char, std::__1::char_traits<char> >&) + 1157
1: 1   libmesh_opt.0.dylib                 0x0000000106b0a55e libMesh::MacroFunctions::report_error(char const*, int, char const*, char const*) + 270
2: 2   libmesh_opt.0.dylib                 0x0000000107165b30 std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, unsigned short, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, unsigned short> > >::mapped_type const& libMesh::Utility::map_find<std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, unsigned short, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, unsigned short> > >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, (std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >*)0>(std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, unsigned short, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, unsigned short> > > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, char const*, int) + 688
3: 3   libmesh_opt.0.dylib                 0x000000010716586c libMesh::System::variable_number(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) const + 28
4: 4   libmesh_opt.0.dylib                 0x0000000106def3fb libMesh::ExodusII_IO::copy_elemental_solution(libMesh::System&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, unsigned int) + 43
5: 5   libmoose-opt.0.dylib                0x0000000105f2ca2d SystemBase::copyVars(libMesh::ExodusII_IO&) + 781
6: 6   libmoose-opt.0.dylib                0x0000000105b524a3 FEProblemBase::initialSetup() + 2531
7: 7   libmoose-opt.0.dylib                0x00000001058be604 Transient::init() + 1764
8: 8   libmoose-opt.0.dylib                0x000000010627198b MooseApp::executeExecutioner() + 75
9: 9   libmoose-opt.0.dylib                0x00000001062721b4 MooseApp::run() + 244
10: 10  sam-opt                             0x0000000104606b7c main + 140
11: 11  libdyld.dylib                       0x00007fff58a453d5 start + 1
12: 12  ???                                 0x0000000000000003 0x0 + 3
[0] ./include/libmesh/utility.h, line 154, compiled Dec 11 2020 at 10:49:02
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
[unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1
:
system msg for write_line failure : Bad file descriptor
```

I'd appreciate it if anyone could give me some suggestions on this issue.
If it is not hard to fix, maybe I can do it.

Please change step-0/1.txt to step-0/1.i
[step-0.txt](https://github.com/idaholab/moose/files/5939900/step-0.txt)
[step-1.txt](https://github.com/idaholab/moose/files/5939901/step-1.txt)

-Ling

**Comment by YaqiWang** [comment, marked as answer]
`v` is the array variable name in MOOSE but not a valid variable name in libMesh. Each component of a moose array variable is a valid libMesh variable with name being `array_variable_name`_# with number being the component index. So the correct syntax in step-1.txt should be either `initial_from_file_var = 'v_0 v_1 v_2'` (possibly not supported yet) or `initial_from_file_var = v` but with some internal name conventions. The later syntax could be better. It needs some code development.

**Reply by GiudGiud** [reply]
@amherm or @smharper might know about array variable restarts.

**Reply by lingzou** [reply]
Thank you @YaqiWang! Yes, I got that there is some internal naming algorithm which causes the issue. I am scoping how much effort that would need to make it work, and if someone (or maybe I myself) can do it.

**Reply by YaqiWang** [reply]
Sorry, I should replied with this. I checked the code, should be fairly strait-forward to support this: in `SystemBase::copyVars`'s `if (hasVariable(vci._dest_name))`, we will need to check `fieldType()` of the obtained moose variable (`SystemBase::isArrayVariable` could be better for this check). If it is array variable, we will need to loop through its components to copy one by one. Array variables have a member function `componentName(unsigned int comp)` to return the libMesh variable name of a component, which should be used in `ExodusII_IO::copy_nodal_solution` or `ExodusII_IO::copy_elemental_solution`.

**Reply by lingzou** [reply]
Thank you @YaqiWang, I can probably give it a try to fix it.

**Reply by lingzou** [reply]
I got it fixed, and will share the code back (moose MR) shortly.

**Reply by YaqiWang** [reply]
Thanks Ling.
